runtime/compiler/z/codegen/J9S390PrivateLinkage.cpp

```diff
@@ -2277,6 +2277,11 @@ TR::S390PrivateLinkage::buildDirectCall(TR::Node * callNode, TR::SymbolReference
       }
 #endif
 
+   // JITServer Workaround: Further transmute dispatchJ9Method symbols to appear as a runtime helper. This will cause OMR to
+   // generate a TR_HelperAddress relocation instead of a TR_RelativeMethodAddress Relocation.
+   if (!comp()->getOption(TR_DisableInliningOfNatives) && callSymbol->getMandatoryRecognizedMethod() == TR::java_lang_invoke_ComputedCalls_dispatchJ9Method)
+      callSymbol->setHelper();
```

Comment:
I think we need an explanation in the commit message for this one. Can you further describe the problem? Also this seems to be only related to JITServer, in which case we should guard it with `if (comp->isOutOfProcessCompilation())` so we can find such places easier in the future and fix them.

Reply:
I do not have any failures to point to for this commit. This change actually comes from a similar change on the x86 side. I encountered this when I initially compared the extra additions JITServer had made to the x86 changes. I decided to include this one as it seemed relevant to Z as well. Here is the original commit, that explains why we set the symbol as a helper here:

https://github.com/eclipse/openj9/commit/fe3c58ddcb69025d48968636810afbd4e4ab6f07

The gist of it is that when the native is inlined, the symref of the call node holds the address of the helper. The code path after we set this property here allows us to call this helper method directly instead of branching to a snippet. However in order to relocate this properly, we need to generate a TR_HelperAddress relocation. That won't happen unless we set the callSymbol's `isHelper` property.

I'm not sure if this is the best solution, or if we should do this here. However I went with it at the time as the workaround is confined to a small enough scope via the `callSymbol->getMandatoryRecognizedMethod() == TR::java_lang_invoke_ComputedCalls_dispatchJ9Method` check.

Given the above, I think this change probably applies to AOT compiles as well.

@mpirvu FYI

Reply:
The change makes sense to me and it may be good to have even in the absence of JITServer. @ymanton since you worked on the relocation infrastructure, what do you think about this one?

Reply:
> I'm not sure if this is the best solution, or if we should do this here.

It does appear to be a workaround for a very specific case for a very specific method. I'm wondering if there are other cases like this. In either way we should apply the same workaround to all codegens or none of them.

Reply:
Looks OK to me.

I can't find any evidence of this in the code right now, but I suspect this inlining of natives optimization is not done in AOT mode. If that's the case that would explain why this is only seen in JITServer mode, but I agree with Marius that this seems like a more general bug.

The change you're making seems like it would be better done where the 'transmutation' first happens:

https://github.com/eclipse/openj9/blob/77b3ec8449d1bbb4c6eecb53e73f56d64441054c/runtime/compiler/env/VMJ9.cpp#L7169-L7179

rather than in every code generator, but perhaps Omer tried this and ran into more issues (since he also had to modify the symref # to get the proper relocation and did so in the above code rather than in the CG).

Alternatively you might be able to fix this closer to home by undoing Omer's change altogether and instead checking for this recognized method in particular somewhere on the relocation creation path (preferably a common place) and creating a helper relocation rather than a relative method relocation, but I don't know if that's better or worse.